# Working with JSON Responses

#### Import dependencies

In [18]:
import requests
import json

#### Perform a GET Request and saving the API's response within a variable

In [21]:
url = "https://api.spacexdata.com/v3/rockets/falcon9"

# Request API response and store in variable called "response"
response = requests.get(url)

# create new variable that contains the json output from the request
response_json = response.json()

print(type(response_json))

<class 'dict'>


#### Pretty print the json output to the screen

In [22]:
print(json.dumps(response_json, indent=4, sort_keys=True))

{
    "active": true,
    "boosters": 0,
    "company": "SpaceX",
    "cost_per_launch": 50000000,
    "country": "United States",
    "description": "Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.",
    "diameter": {
        "feet": 12,
        "meters": 3.7
    },
    "engines": {
        "engine_loss_max": 2,
        "layout": "octaweb",
        "number": 9,
        "propellant_1": "liquid oxygen",
        "propellant_2": "RP-1 kerosene",
        "thrust_sea_level": {
            "kN": 845,
            "lbf": 190000
        },
        "thrust_to_weight": 180.1,
        "thrust_vacuum": {
            "kN": 914,
            "lbf": 205500
        },
        "type": "merlin",
        "version": "1D+"
    },
    "first_flight": "2010-06-04",
    "first_stage": {
        "burn_time_sec": 162,
        "engines": 9,
        "fuel_amount_tons": 385,
        "reusable": true,
        "t

#### Since we are now working with a dictionary, it is possible to grab a specific value from within the JSON object

In [28]:
print(response_json["payload_weights"])

[{'id': 'leo', 'name': 'Low Earth Orbit', 'kg': 22800, 'lb': 50265}, {'id': 'gto', 'name': 'Geosynchronous Transfer Orbit', 'kg': 8300, 'lb': 18300}, {'id': 'mars', 'name': 'Mars Orbit', 'kg': 4020, 'lb': 8860}]


#### It is also possible to perform some analyses on values stored within the JSON object

In [29]:
number_payloads = len(response_json["payload_weights"])
print(f"There are {number_payloads} payloads.")

There are 3 payloads.


#### Finally, it's possible to reference the values stored within sub-dictionaries and sub-lists

In the example below, we're querying the **dictionary** where the key is `payload_weights`. The value for this key is a list of more dictionaries. We're then asking for the first (index 0) item in that list, and then asking for the value for the "`kg`" key. See below for the `payload_weights` dictionary. Use this to trace the code below to the value that we're retrieving.

`"payload_weights": [
        {
            "id": "leo",
            "kg": 22800,
            "lb": 50265,
            "name": "Low Earth Orbit"
        },
        {
            "id": "gto",
            "kg": 8300,
            "lb": 18300,
            "name": "Geosynchronous Transfer Orbit"
        },
        {
            "id": "mars",
            "kg": 4020,
            "lb": 8860,
            "name": "Mars Orbit"
        }
    ]`

In [30]:
payload_weight = response_json["payload_weights"][0]["kg"]
print(f"The first payload weighed {payload_weight} Kilograms")

The first payload weighed 22800 Kilograms
